contracts/governance/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+#[repr(u32)]
+pub enum Error {
+    /// The contract should be only initialized once.
+    AlreadyInitialized = 1,
+    // Must have funds for the operation.
+    InsufficientFunds = 2,
+    // Certain amounts are not valid in some operations.(Like under and/or equal to zero)
+    InvalidAmount = 3,
+    ParticipantNotFound = 4,
+    ParticipantNotWhitelisted = 5,
+    ExpectedStorageKeyNotFound = 6,
+    ProposalNeedsApproval = 7,
+    OnlyAuthorCanExecuteProposals = 8,
+    AlreadyExecuted = 9,
+}
+
+#[contract]
+pub struct GovernanceContract;
+
+#[contractimpl]
+impl GovernanceContract {
+    pub fn init(
+        env: Env,
+        curator: Address,
+        token: Address,
+        voting_period_secs: u64,
+        target_approval_rate_bps: u32,
+        salt: BytesN<32>,
+    ) {
+        let storage = env.storage().persistent();
+
+        if storage.has(&DataKey::Initialized) {
+            panic_with_error!(&env, Error::AlreadyInitialized)
+        }
+
+        // Deploy the voting contract (A dependency of this one)
+        let voting_contract_hash = env.deployer().upload_contract_wasm(voting_contract::WASM);
+        let deployer = env.deployer().with_current_contract(salt);
+        let voting_contract_address = deployer.deploy(voting_contract_hash);
+        let voting_client = voting_contract::Client::new(&env, &voting_contract_address);
+
+        // Init the voting contract.
+        voting_client.init(
+            &env.current_contract_address(),
+            &voting_period_secs,
+            &target_approval_rate_bps,
+            &u128::MAX, // This is a dummy value as participation state will be managed by this contract due to data locality.  It needs to be positive.
+            &true,      // Only this contract can do operation on behalf of the participants.
+        );
+
+        env.events().publish(
+            (Symbol::new(&env, "voting_contract_initialized"),),
+            voting_contract_address.clone(),
+        );
+
+        storage.set(&DataKey::Initialized, &());
+        storage.set(&DataKey::Curator, &curator);
+        storage.set(&DataKey::Token, &token);
+        storage.set(&DataKey::VotingContractAddress, &voting_contract_address);
+        storage.set(
+            &DataKey::Participants,
+            &Map::<Address, Participant>::new(&env),
+        );
+        storage.set(&DataKey::ExecutedProposals, &Map::<u64, ()>::new(&env));
+    }
+
+    pub fn join(env: Env, participant_addr: Address, amount: i128) -> Result<(), Error> {
+        participant_addr.require_auth();
+
+        let storage = env.storage().persistent();
+        let mut participant_repo = participant::Repository::new(&storage)?;
+        let mut participant = Participant::new(participant_addr.clone());
+
+        Self::stake_funds(&env, &mut participant, amount)?;
+
+        participant_repo.save(participant);
+
+        env.events().publish(
+            (Symbol::new(&env, "participant_joined"), participant_addr),
+            (),
```

Comment:
Is this needed to be unit?
Can we emit like the current count of participants?

Reply:
Fixed at eedfd82663a54ceac2c78cc0751c5f67c66259ff